In [85]:
import numpy as np
import pandas as pd

In [86]:
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

In [87]:
ls

 Volume in drive C is Windows X-Lite
 Volume Serial Number is FAAD-6DE0

 Directory of C:\Users\Admin\sandbox\Portfolio\lost sales\retail store data

07/27/2026  10:36 PM    <DIR>          .
07/26/2026  06:19 PM    <DIR>          ..
07/27/2026  10:36 PM         6,190,180 cleaned_inventory.csv
11/24/2024  08:09 PM         6,191,463 retail_store_inventory.csv
               2 File(s)     12,381,643 bytes
               2 Dir(s)  132,517,027,840 bytes free


In [88]:
cd C:\Users\Admin\sandbox\Portfolio\lost sales\retail store data

C:\Users\Admin\sandbox\Portfolio\lost sales\retail store data


In [89]:
df = pd.read_csv("retail_store_inventory.csv")

In [90]:
df.head()

,Date,Store ID,Product ID,Category,Region,Inventory Level,Units Sold,Units Ordered,Demand Forecast,Price,Discount,Weather Condition,Holiday/Promotion,Competitor Pricing,Seasonality
0,2022-01-01,S001,P0001,Groceries,North,231,127,55,135.47,33.50,20,Rainy,0,29.69,Autumn
1,2022-01-01,S001,P0002,Toys,South,204,150,66,144.04,63.01,20,Sunny,0,66.16,Autumn
2,2022-01-01,S001,P0003,Toys,West,102,65,51,74.02,27.99,10,Sunny,1,31.32,Summer
3,2022-01-01,S001,P0004,Toys,North,469,61,164,62.18,32.72,10,Cloudy,1,34.74,Autumn
4,2022-01-01,S001,P0005,Electronics,East,166,14,135,9.26,73.64,0,Sunny,0,68.95,Summer


In [91]:
df.shape

(73100, 15)

In [92]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 73100 entries, 0 to 73099
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Date                73100 non-null  str    
 1   Store ID            73100 non-null  str    
 2   Product ID          73100 non-null  str    
 3   Category            73100 non-null  str    
 4   Region              73100 non-null  str    
 5   Inventory Level     73100 non-null  int64  
 6   Units Sold          73100 non-null  int64  
 7   Units Ordered       73100 non-null  int64  
 8   Demand Forecast     73100 non-null  float64
 9   Price               73100 non-null  float64
 10  Discount            73100 non-null  int64  
 11  Weather Condition   73100 non-null  str    
 12  Holiday/Promotion   73100 non-null  int64  
 13  Competitor Pricing  73100 non-null  float64
 14  Seasonality         73100 non-null  str    
dtypes: float64(3), int64(5), str(7)
memory usage: 8.4 MB


In [93]:
df.columns

Index(['Date', 'Store ID', 'Product ID', 'Category', 'Region',
       'Inventory Level', 'Units Sold', 'Units Ordered', 'Demand Forecast',
       'Price', 'Discount', 'Weather Condition', 'Holiday/Promotion',
       'Competitor Pricing', 'Seasonality'],
      dtype='str')

Key findings:
1. here i can see that overall we dont have any explicit null values 
2. also the date column is stores as str which should be datetime
3. also here in holiday clm its property must be as boolean.

data quality checking and assesment

In [94]:
df.isnull().sum()

Date                  0
Store ID              0
Product ID            0
Category              0
Region                0
Inventory Level       0
Units Sold            0
Units Ordered         0
Demand Forecast       0
Price                 0
Discount              0
Weather Condition     0
Holiday/Promotion     0
Competitor Pricing    0
Seasonality           0
dtype: int64

In [95]:
df.duplicated().sum()

np.int64(0)

In [96]:
df.duplicated(subset=["Date", "Store ID", "Product ID"]).sum()

np.int64(0)

key observation on data quality check
1. we do not have any null values overall
2. no duplicate values in data too

In [97]:
df["Store ID"].value_counts()

Store ID
S001    14620
S002    14620
S003    14620
S004    14620
S005    14620
Name: count, dtype: int64

In [98]:
df["Product ID"].value_counts()

Product ID
P0001    3655
P0002    3655
P0003    3655
P0004    3655
P0005    3655
P0006    3655
P0007    3655
P0008    3655
P0009    3655
P0010    3655
P0011    3655
P0012    3655
P0013    3655
P0014    3655
P0015    3655
P0016    3655
P0017    3655
P0018    3655
P0019    3655
P0020    3655
Name: count, dtype: int64

In [99]:
df["Category"].value_counts()

Category
Furniture      14699
Toys           14643
Clothing       14626
Groceries      14611
Electronics    14521
Name: count, dtype: int64

so overall we found out that we have uniuqe
1.stores - 5
2. products - 20
3. category - 5

In [100]:
df["Date"] = pd.to_datetime(df["Date"])

In [101]:
df["Date"].dtype

dtype('<M8[us]')

In [102]:
#date range 
print("Start Date :", df["Date"].min())
print("End Date   :", df["Date"].max())

Start Date : 2022-01-01 00:00:00
End Date   : 2024-01-01 00:00:00


In [103]:
 df["Date"].nunique()

731

in conlcusion we have 2 years of date 

In [104]:
df.describe()

,Date,Inventory Level,Units Sold,Units Ordered,Demand Forecast,Price,Discount,Holiday/Promotion,Competitor Pricing
count,73100,73100.000000,73100.000000,73100.000000,73100.000000,73100.000000,73100.000000,73100.000000,73100.000000
mean,2023-01-01 00:00:00,274.469877,136.464870,110.004473,141.494720,55.135108,10.009508,0.497305,55.146077
min,2022-01-01 00:00:00,50.000000,0.000000,20.000000,-9.990000,10.000000,0.000000,0.000000,5.030000
25%,2022-07-02 00:00:00,162.000000,49.000000,65.000000,53.670000,32.650000,5.000000,0.000000,32.680000
50%,2023-01-01 00:00:00,273.000000,107.000000,110.000000,113.015000,55.050000,10.000000,0.000000,55.010000
75%,2023-07-03 00:00:00,387.000000,203.000000,155.000000,208.052500,77.860000,15.000000,1.000000,77.820000
max,2024-01-01 00:00:00,500.000000,499.000000,200.000000,518.550000,100.000000,20.000000,1.000000,104.940000
std,NaN,129.949514,108.919406,52.277448,109.254076,26.021945,7.083746,0.499996,26.191408


In [105]:
negative_demand = df[df["Demand Forecast"] < 0]
print("Negative Demand Forecast values found :", len(negative_demand))

Negative Forecast Records : 673


### Data Quality Finding — Negative Demand Forecast Values

We identified **673 negative values** in the `Demand Forecast` column. Since a demand forecast cannot logically be negative, these values are treated as invalid and corrected during the data-cleaning stage by replacing them with **0**.

This ensures the cleaned dataset contains only valid non-negative demand forecast values.

In [106]:
#check the discount validation as compared to the selling price
df["Discount"].describe()

count    73100.000000
mean        10.009508
std          7.083746
min          0.000000
25%          5.000000
50%         10.000000
75%         15.000000
max         20.000000
Name: Discount, dtype: float64

so here discount values are between 0% to 20%, and there are no values outside this range.

In [107]:
numeric_columns = [
    "Inventory Level",
    "Units Sold",
    "Units Ordered",
    "Demand Forecast",
    "Price",
    "Discount",
    "Competitor Pricing"
]

df[numeric_columns].max()

Inventory Level       500.00
Units Sold            499.00
Units Ordered         200.00
Demand Forecast       518.55
Price                 100.00
Discount               20.00
Competitor Pricing    104.94
dtype: float64

In [108]:
df["Holiday/Promotion"].value_counts()

Holiday/Promotion
0    36747
1    36353
Name: count, dtype: int64

data quality have been cheked now its time for data cleaning for further process

## DATA CLEANING

In [109]:
# Clean negative Demand Forecast values
negative_forecast = (df["Demand Forecast"] < 0).sum()
print("Negative Demand Forecast values found:", negative_forecast)

673


In [110]:
# Replace all negative Demand Forecast values with 0
df.loc[df["Demand Forecast"] < 0, "Demand Forecast"] = 0

In [111]:
# Verify that no negative Demand Forecast values remain
remaining_negative = (df["Demand Forecast"] < 0).sum()
print("Negative values remaining after cleaning:", remaining_negative)

negative values : 0


In [112]:
df.to_csv("cleaned_inventory.csv", index=False)

print("Cleaned dataset saved")

Cleaned dataset saved


In [113]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 73100 entries, 0 to 73099
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Date                73100 non-null  datetime64[us]
 1   Store ID            73100 non-null  str           
 2   Product ID          73100 non-null  str           
 3   Category            73100 non-null  str           
 4   Region              73100 non-null  str           
 5   Inventory Level     73100 non-null  int64         
 6   Units Sold          73100 non-null  int64         
 7   Units Ordered       73100 non-null  int64         
 8   Demand Forecast     73100 non-null  float64       
 9   Price               73100 non-null  float64       
 10  Discount            73100 non-null  int64         
 11  Weather Condition   73100 non-null  str           
 12  Holiday/Promotion   73100 non-null  int64         
 13  Competitor Pricing  73100 non-null  float64       
 14  S

In [ ]:
df